packages/phone-number-privacy/src/index.ts

```diff
@@ -1,21 +1,148 @@
   try {
-    // Adding this here as an example of how to connect to the DB
-    const knex = connectToDatabase()
-    knex('accounts')
-      .first()
-      .then((val) => console.debug('account data', val))
-      .catch((reason) => console.error(reason))
-
-    const salt = BLINDBLS.computePRF(privateKey, new Buffer(request.body.blindPhoneNumber))
-    response.json({ success: true, salt })
+    const privateKey = new Buffer(config.salt.key)
+    return BLINDBLS.computePRF(privateKey, new Buffer(queryPhoneNumber))
   } catch (e) {
-    console.error('Failed to compute BLS salt', e)
-    response.status(500).send('Failed to compute BLS salt')
+    console.error('Failed to compute salt', e)
+    response.status(500).send('Failed to compute salt')
   }
-})
+}
+
+/*
+ * Returns how many queries the account can make based on the
+ * calculated query quota and the number of queries already performed.
+ */
+async function getRemainingQueryCount(account: string, phoneNumber: string) {
+  const queryQuota = await getQueryQuota(account, phoneNumber)
+  const performedQueryCount = await getPerformedQueryCount(account)
+  return queryQuota - performedQueryCount
+}
+
+/*
+ * Calculates how many queries the caller has unlocked based on the algorithm
+ * unverifiedQueryCount + verifiedQueryCount + (queryPerTransaction * transactionCount)
+ */
+async function getQueryQuota(account: string, phoneNumber: string) {
+  let queryQuota: number = config.salt.unverifiedQueryCount
+  if (await isVerified(account, phoneNumber)) {
+    queryQuota += config.salt.verifiedQueryCount
+    const transactionCount = await getTransactionCountFromAccount(account)
+    queryQuota += config.salt.queryPerTransaction * transactionCount
+  }
+  return queryQuota
+}
+
+async function isVerified(account: string, phoneNumber: string): Promise<boolean> {
+  const attestationsWrapper: AttestationsWrapper = await contractKit.contracts.getAttestations()
+  const attestationStats = await attestationsWrapper.getAttestationStat(phoneNumber, account)
+  const numAttestationsCompleted = attestationStats.completed
+  const numAttestationsRemaining = NUM_ATTESTATIONS_REQUIRED - numAttestationsCompleted
+  return numAttestationsRemaining <= 0
+}
+
+async function getTransactionCountFromAccount(account: string): Promise<number> {
+  // TODO [amyslawson] consider adding retry
+  const transactionUrl = `${config.blockchain.blockscout}/api?module=account&action=tokentx&address=${account}`
+  const resp = await fetch(transactionUrl)
+  const jsonResp = await resp.json()
+  if (jsonResp == null || jsonResp.result == null) {
+    console.error(`Failed to get valid response for ${transactionUrl}`)
+    return 0
+  }
+  return jsonResp.result.filter((transaction: any) => transaction.from === account).length
+}
+
+/*
+ * Returns how many queries the account has already performed.
+ */
+async function getPerformedQueryCount(account: string): Promise<number> {
```

Comment:
We should create a new class for wrapping the knex object and performing these high level operations. The more internals we expose, the harder it will be to guarantee our assumptions. Having a separate class will also make it easier to mock during testing. I recommend creating a new class (ex. `QueryStateStore`) that exposes `getPerformedQueryCount` and `incrementQueryCount`.